_posts/2022-08-10-architecting-trino-next-gen-etl-engine.md

```diff
@@ -0,0 +1,171 @@
+from different sources. However, after seeing the merit of a unified engine for
+both interactive and batch workloads, it quickly got popular as an ELT/ETL
+engine in the industry. In practice though, insufficient support for
+memory-intensive and long-running queries has been major pain points that
+prevent users from adopting Trino for batch processing at larger scale.
+Inability to recover from failures in long-running queries leads to huge
+variance of data landing time, and lack of fine-grained resource management
+requires all data of stage execution to fit in memory. This blog discusses the
+architecture we used to address these pain points and provide first-class
+support for large-scale batch processing in Trino.
+
+<!--more-->
+
+# Limitations of the original architecture
+
+![](/assets/blog/architecting-trino-next-gen-etl-engine/previous-architecture.png)
+
+Trino is designed for low latency and follows the classic MPP architecture,
+featuring in-memory streaming shuffle. Tasks are interconnected and data is
+streamed through the task dependency graph with no intermediate checkpointing.
+This all-or-nothing architecture makes it really hard to tolerate faults. If
+any of the tasks dies/fails, there is no way to resume execution from some
+intermediate state, the only option is to restart the whole query. This is
+extremely painful for long-running queries as they are more likely to fail, and
+impact of failures is much higher as it often results in significant waste of
+time and resources.
+
+With the exception of grouped execution,which applies to limited scenarios like
+bucketed tables in Hive, Trino requires processing of all partitions in a stage
+to happen at the same time. As a result, data processing of all these partitions
+must fit in memory. This memory barrier constraint makes it extremely difficult
+to execute huge joins/aggregations in ETL pipelines, when the cluster doesn’t
+have enough memory to hold the hash tables.
+
+Using static task placement, we are limited by the size of the cluster in how
+many nodes we can partition the data. With inflexible scheduling runtime,
+certain optimizations are not possible. We can’t apply classic techniques like resource
+isolation, speculative execution, adaptive query planning, and skew handling to
+scale Trino worker execution.
+
+# A new execution paradigm
+
+![](/assets/blog/architecting-trino-next-gen-etl-engine/current-architecture.png)
+
+To address this, we introduce a new execution paradigm to Trino by adding
+exchange spooling that stores intermediate data in an efficient buffering layer
+at stage boundaries. This opens up a lot of new opportunities around failure
+recovery, resource management, and query optimization.
+
+## Fault-tolerance
+
+There are now two levels of fault-tolerance into Trino:
+
+### Query-level Retry
+
+By maintaining a deduplication buffer with the ability to spill to external
```

Comment:
"deduplication buffer" needs some clarification. It's not obvious what that means for someone unfamiliar with the implementation